### PR TITLE
[TR] PIM-4052: add current user in versioning

### PR DESCRIPTION
**Pull request by skeleton**
| Q                    | A
| -------------------- | ---
| Bug fix?             | N
| New feature?         | Y
| BC breaks?           | N
| CI currently passes? | N
| Tests pass?          | Y
| Scenarios pass?      | Y
| Checkstyle issues?*  | N
| PMD issues?**        | N
| Changelog updated?   | N
| Fixed tickets        | PIM-4052
| DB schema updated?   | N 
| Migration script?    | N
| Doc PR               | N

### CODE REVIEW

File: src/Pim/Bundle/VersioningBundle/Manager/VersionManager.php [hunk 1]
```diff
@@ -112,6 +117,11 @@ public function buildVersion($versionable, array $changeset = [])
     {
         $createdVersions = [];
 
+        $event = $this->eventDispatcher->dispatch(UsernameEvents::FIND_USERNAME, new UsernameEvent());
```
Comment (jjanvier) [inline src/Pim/Bundle/VersioningBundle/Manager/VersionManager.php]: I understand why you do that, but I disagree on the name of the event.

Here you "pre build a version", and something should subscribe to this event to retrieve the username

File: src/Pim/Bundle/VersioningBundle/EventSubscriber/AddUserSubscriber.php [hunk 2]
```diff
@@ -17,23 +16,14 @@
...
-     * @var VersionManager
-     */
-    protected $versionManager;
-
-    /**
-     * @param VersionManager           $versionManager
      * @param SecurityContextInterface $securityContext
      */
-    public function __construct(VersionManager $versionManager, SecurityContextInterface $securityContext = null)
+    public function __construct(SecurityContextInterface $securityContext = null)
```
Comment (jjanvier) [inline src/Pim/Bundle/VersioningBundle/EventSubscriber/AddUserSubscriber.php]: add this to the changelog as a BC

File: src/Pim/Bundle/VersioningBundle/EventSubscriber/AddUserSubscriber.php [hunk 3]
```diff
@@ -42,23 +32,27 @@ public function __construct(VersionManager $versionManager, SecurityContextInter
...
     }
 
     /**
-     * @param GetResponseEvent $event
+     * @param UsernameEvent $event
+     *
+     * @return UsernameEvent
      */
-    public function onKernelRequest(GetResponseEvent $event)
+    public function findUsername(UsernameEvent $event)
```
Comment (jjanvier) [inline src/Pim/Bundle/VersioningBundle/EventSubscriber/AddUserSubscriber.php]: BC breaks

File: spec/Pim/Bundle/VersioningBundle/Manager/VersionManagerSpec.php [hunk 4]
```diff
@@ -11,6 +11,8 @@
 use Pim\Bundle\VersioningBundle\Model\Version;
 use Pim\Bundle\VersioningBundle\Repository\VersionRepositoryInterface;
 use Prophecy\Argument;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Symfony\Component\Security\Core\SecurityContextInterface;
```
Comment (grena) [inline spec/Pim/Bundle/VersioningBundle/Manager/VersionManagerSpec.php]: unused

File: src/Pim/Bundle/VersioningBundle/Event/PreBuildVersionEvent.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ */
+class PreBuildVersionEvent extends Event
+{
+    /** @var string */
+    protected $username;
+
+    /**
+     * @param string $username
+     *
+     * @return UsernameEvent
```
Comment (willy-ahva) [inline src/Pim/Bundle/VersioningBundle/Event/PreBuildVersionEvent.php]: @return PreBuildVersionEvent

File: src/Pim/Bundle/VersioningBundle/Event/PreBuildVersionEvents.php [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+ * @author    Marie Bochu <marie.bochu@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class PreBuildVersionEvents
+{
+    /**
+     * This event is dispatched before build a new version
+     *
+     * The event listener receives an
```
Comment (willy-ahva) [inline src/Pim/Bundle/VersioningBundle/Event/PreBuildVersionEvents.php]: receives "a"

File: CHANGELOG.md [hunk 7]
```diff
@@ -16,6 +16,9 @@
 - PIM-4024: Fix for metric and price denormalizer
 
 ## BC breaks
+- Change the constructor of `Pim\Bundle\VersioningBundle\EventSubscriber\AddUserSubscriber`, removed `Pim\Bundle\VersioningBundle\Manager\VersionManager`
+- Rename method `onKernelRequest` to `findUsername` on `Pim\Bundle\VersioningBundle\EventSubscriber\AddUserSubscriber`
+- Change the constructor of `Pim\Bundle\VersioningBundle\Manager\VersionManager`, added `Symfony\Component\EventDispatcher\EventDispatcherInterface` as the last argument
```
Comment (jmleroux) [inline CHANGELOG.md]: Do we need to precise that this is the last argument ?
If we ever modify the constructor again, this precision will not be clear.

**Pull request merged**